File: src/tool.rs
use std::collections::HashMap;
use std::future::Future;
use std::pin::Pin;

use serde::de::DeserializeOwned;
use serde_json::Value;

use crate::{Error, Result};

// ---------------------------------------------------------------------------
// Tool trait
// ---------------------------------------------------------------------------

/// A tool that can be called by an LLM agent.
///
/// Shared between motosan-chat (chat bots) and crucible-agent (forum pipelines).
pub trait Tool: Send + Sync {
    /// Return the tool definition (name, description, input schema).
    fn def(&self) -> ToolDef;

    /// Execute the tool with the given arguments and context.
    fn call(
        &self,
        args: Value,
        ctx: &ToolContext,
    ) -> Pin<Box<dyn Future<Output = ToolResult> + Send + '_>>;
}

// ---------------------------------------------------------------------------
// ToolDef
// ---------------------------------------------------------------------------

/// Definition of a tool, suitable for serialization to LLM APIs (Claude, OpenAI, etc.).
#[derive(Debug, Clone, PartialEq)]
pub struct ToolDef {
    pub name: String,
    pub description: String,
    pub input_schema: Value,
}

impl ToolDef {
    /// Validate that the input_schema itself is well-formed.
    pub fn validate_input_schema(&self) -> Result<()> {
        let schema = self
            .input_schema
            .as_object()
            .ok_or_else(|| Error::Validation("input_schema must be a JSON object".into()))?;

        if schema.get("type").and_then(Value::as_str) != Some("object") {
            return Err(Error::Validation(
                "input_schema.type must be \"object\"".into(),
            ));
        }

        let properties = schema
            .get("properties")
            .and_then(Value::as_object)
            .ok_or_else(|| {
                Error::Validation("input_schema.properties must be an object".into())
            })?;

        if let Some(required) = schema.get("required") {
            let required = required
                .as_array()
                .ok_or_else(|| Error::Validation("input_schema.required must be an array".into()))?;

            for field in required {
                let field_name = field.as_str().ok_or_else(|| {
                    Error::Validation("input_schema.required entries must be strings".into())
                })?;
                if !properties.contains_key(field_name) {
                    return Err(Error::Validation(format!(
                        "required field not in properties: {field_name}"
                    )));
                }
            }
        }

        Ok(())
    }

    /// Validate arguments against the input_schema (type checking + enum checking).
    pub fn validate_args(&self, args: &Value) -> Result<()> {
        self.validate_input_schema()?;

        let schema = self
            .input_schema
            .as_object()
            .ok_or_else(|| Error::Validation("input_schema must be a JSON object".into()))?;
        let properties = schema
            .get("properties")
            .and_then(Value::as_object)
            .ok_or_else(|| {
                Error::Validation("input_schema.properties must be an object".into())
            })?;

        let args = args
            .as_object()
            .ok_or_else(|| Error::Validation("tool args must be a JSON object".into()))?;

        // Check required fields
        if let Some(required) = schema.get("required").and_then(Value::as_array) {
            for field in required {
                let field_name = field
                    .as_str()
                    .ok_or_else(|| Error::Validation("required field name must be string".into()))?;
                if !args.contains_key(field_name) {
                    return Err(Error::MissingField(field_name.into()));
                }
            }
        }

        // Type and enum checking
        for (key, value) in args {
            let Some(spec) = properties.get(key).and_then(Value::as_object) else {
                continue;
            };

            if let Some(expected_type) = spec.get("type").and_then(Value::as_str) {
                let type_matches = match expected_type {
                    "string" => value.is_string(),
                    "number" => value.is_number(),
                    "integer" => value.as_i64().is_some() || value.as_u64().is_some(),
                    "boolean" => value.is_boolean(),
                    "object" => value.is_object(),
                    "array" => value.is_array(),
                    "null" => value.is_null(),
                    _ => true,
                };

                if !type_matches {
                    return Err(Error::Validation(format!(
                        "field {key} expected type {expected_type}"
                    )));
                }
            }

            if let Some(enum_values) = spec.get("enum").and_then(Value::as_array) {
                if !enum_values.contains(value) {
                    return Err(Error::Validation(format!("field {key} is not in enum")));
                }
            }
        }

        Ok(())
    }

    /// Validate and deserialize args into a typed struct.
    pub fn parse_args<T: DeserializeOwned>(&self, args: Value) -> Result<T> {
        self.validate_args(&args)?;
        serde_json::from_value(args).map_err(|err| Error::Parse(format!("invalid args: {err}")))
    }
}

// ---------------------------------------------------------------------------
// ToolResult
// ---------------------------------------------------------------------------

/// Result returned by tool execution.
///
/// Combines motosan-chat's typed content with crucible-agent's metadata fields.
#[derive(Debug, Clone)]
pub struct ToolResult {
    /// Structured content blocks returned by the tool.
    pub content: Vec<ToolContent>,
    /// Whether the result represents an error.
    pub is_error: bool,
    /// Source URL for citation (e.g. web_search result, fetch_url target).
    pub citation: Option<String>,
    /// Whether this result should be injected into the next round's context.
    pub inject_to_context: bool,
    /// Execution time in milliseconds.
    pub duration_ms: Option<u64>,
}

impl ToolResult {
    /// Successful text result.
    pub fn text(text: impl Into<String>) -> Self {
        Self {
            content: vec![ToolContent::Text(text.into())],
            is_error: false,
            citation: None,
            inject_to_context: false,
            duration_ms: None,
        }
    }

    /// Successful JSON result.
    pub fn json(value: Value) -> Self {
        Self {
            content: vec![ToolContent::Json(value)],
            is_error: false,
            citation: None,
            inject_to_context: false,
            duration_ms: None,
        }
    }

    /// Error result.
    pub fn error(message: impl Into<String>) -> Self {
        Self {
            content: vec![ToolContent::Text(message.into())],
            is_error: true,
            citation: None,
            inject_to_context: false,
            duration_ms: None,
        }
    }

    /// Set citation on this result (builder pattern).
    pub fn with_citation(mut self, citation: impl Into<String>) -> Self {
        self.citation = Some(citation.into());
        self
    }

    /// Set inject_to_context on this result.
    pub fn with_inject(mut self, inject: bool) -> Self {
        self.inject_to_context = inject;
        self
    }

    /// Set duration_ms on this result.
    pub fn with_duration(mut self, ms: u64) -> Self {
        self.duration_ms = Some(ms);
        self
    }

    /// Get the first text content, if any.
    pub fn as_text(&self) -> Option<&str> {
        self.content.iter().find_map(|c| match c {
            ToolContent::Text(s) => Some(s.as_str()),
            _ => None,
        })
    }
}

// ---------------------------------------------------------------------------
// ToolContent
// ---------------------------------------------------------------------------

/// A single content block in a tool result.
#[derive(Debug, Clone, PartialEq)]
pub enum ToolContent {
    /// Plain text content.
    Text(String),
    /// Structured JSON content.
    Json(Value),
}

// ---------------------------------------------------------------------------
// ToolContext
// ---------------------------------------------------------------------------

/// Execution context passed to every tool call.
///
/// Contains common fields shared across platforms, plus an `extra` map for
/// platform-specific data (crucible: org_id, project_id; chat: group_id, etc.).
#[derive(Debug, Clone, Default)]
pub struct ToolContext {
    /// Who is calling the tool (agent_id or user_id).
    pub caller_id: String,
    /// Platform identifier ("crucible", "line", "telegram", "discord", etc.).
    pub platform: String,
    /// Platform-specific key-value extensions.
    pub extra: HashMap<String, Value>,
}

impl ToolContext {
    pub fn new(caller_id: impl Into<String>, platform: impl Into<String>) -> Self {
        Self {
            caller_id: caller_id.into(),
            platform: platform.into(),
            extra: HashMap::new(),
        }
    }

    /// Insert an extra field (builder pattern).
    pub fn with(mut self, key: impl Into<String>, value: impl Into<Value>) -> Self {
        self.extra.insert(key.into(), value.into());
        self
    }

    /// Get a string value from extra.
    pub fn get_str(&self, key: &str) -> Option<&str> {
        self.extra.get(key)?.as_str()
    }

    /// Get a u64 value from extra.
    pub fn get_u64(&self, key: &str) -> Option<u64> {
        self.extra.get(key)?.as_u64()
    }

    /// Get a bool value from extra.
    pub fn get_bool(&self, key: &str) -> Option<bool> {
        self.extra.get(key)?.as_bool()
    }
}

// ---------------------------------------------------------------------------
// Tests
// ---------------------------------------------------------------------------

#[cfg(test)]
mod tests {
    use super::*;
    use serde::Deserialize;
    use serde_json::json;

    #[derive(Debug, Deserialize)]
    struct SearchArgs {
        query: String,
        max_results: Option<u32>,
    }

    fn search_def() -> ToolDef {
        ToolDef {
            name: "web_search".into(),
            description: "Search the web".into(),
            input_schema: json!({
                "type": "object",
                "properties": {
                    "query": { "type": "string" },
                    "max_results": { "type": "integer" }
                },
                "required": ["query"]
            }),
        }
    }

    #[test]
    fn validate_input_schema_accepts_valid() {
        search_def().validate_input_schema().unwrap();
    }

    #[test]
    fn validate_input_schema_rejects_missing_properties() {
        let def = ToolDef {
            name: "bad".into(),
            description: "bad".into(),
            input_schema: json!({ "type": "object" }),
        };
        assert!(def.validate_input_schema().is_err());
    }

    #[test]
    fn validate_args_accepts_valid() {
        let def = search_def();
        let args = json!({ "query": "rust" });
        def.validate_args(&args).unwrap();
    }

    #[test]
    fn validate_args_rejects_missing_required() {
        let def = search_def();
        let args = json!({ "max_results": 5 });
        assert!(matches!(
            def.validate_args(&args),
            Err(crate::Error::MissingField(_))
        ));
    }

    #[test]
    fn validate_args_rejects_wrong_type() {
        let def = search_def();
        let args = json!({ "query": 123 });
        assert!(def.validate_args(&args).is_err());
    }

    #[test]
    fn validate_args_checks_enum() {
        let def = ToolDef {
            name: "t".into(),
            description: "t".into(),
            input_schema: json!({
                "type": "object",
                "properties": {
                    "lang": { "type": "string", "enum": ["en", "ja", "zh"] }
                },
                "required": ["lang"]
            }),
        };
        def.validate_args(&json!({ "lang": "en" })).unwrap();
        assert!(def.validate_args(&json!({ "lang": "fr" })).is_err());
    }

    #[test]
    fn parse_args_deserializes_typed_struct() {
        let def = search_def();
        let args = json!({ "query": "rust", "max_results": 10 });
        let parsed: SearchArgs = def.parse_args(args).unwrap();
        assert_eq!(parsed.query, "rust");
        assert_eq!(parsed.max_results, Some(10));
    }

    #[test]
    fn tool_result_text_roundtrip() {
        let r = ToolResult::text("hello");
        assert!(!r.is_error);
        assert_eq!(r.as_text(), Some("hello"));
        assert!(r.citation.is_none());
    }

    #[test]
    fn tool_result_error_sets_flag() {
        let r = ToolResult::error("boom");
        assert!(r.is_error);
    }

    #[test]
    fn tool_result_builder_chain() {
        let r = ToolResult::text("data")
            .with_citation("https://example.com")
            .with_inject(true)
            .with_duration(42);
        assert_eq!(r.citation.as_deref(), Some("https://example.com"));
        assert!(r.inject_to_context);
        assert_eq!(r.duration_ms, Some(42));
    }

    #[test]
    fn tool_context_extra_helpers() {
        let ctx = ToolContext::new("agent-1", "crucible")
            .with("org_id", json!("motosan"))
            .with("budget", json!(5));
        assert_eq!(ctx.get_str("org_id"), Some("motosan"));
        assert_eq!(ctx.get_u64("budget"), Some(5));
        assert_eq!(ctx.get_str("missing"), None);
    }
}

File: typescript/src/tool.ts
import { ToolError } from "./error.js";

// ---------------------------------------------------------------------------
// ToolContent
// ---------------------------------------------------------------------------

export interface TextContent {
  type: "text";
  data: string;
}

export interface JsonContent {
  type: "json";
  data: unknown;
}

export type ToolContent = TextContent | JsonContent;

// ---------------------------------------------------------------------------
// ToolDef
// ---------------------------------------------------------------------------

export class ToolDef {
  readonly name: string;
  readonly description: string;
  readonly inputSchema: Record<string, unknown>;

  constructor(
    name: string,
    description: string,
    inputSchema: Record<string, unknown>,
  ) {
    this.name = name;
    this.description = description;
    this.inputSchema = inputSchema;
  }

  /**
   * Validates that the input schema itself is structurally sound.
   * Throws a ToolError if the schema is invalid.
   */
  validateInputSchema(): void {
    if (!this.name || this.name.trim().length === 0) {
      throw ToolError.missingField("name");
    }
    if (!this.description || this.description.trim().length === 0) {
      throw ToolError.missingField("description");
    }
    if (this.inputSchema === null || this.inputSchema === undefined) {
      throw ToolError.missingField("inputSchema");
    }
    if (typeof this.inputSchema !== "object" || Array.isArray(this.inputSchema)) {
      throw ToolError.validation("inputSchema must be a plain object");
    }
  }

  /**
   * Validates that the given args conform to the input schema.
   * Performs basic structural validation (type check for "object" schemas
   * and required-field checking).
   */
  validateArgs(args: unknown): void {
    const schemaType = this.inputSchema["type"];

    if (schemaType === "object") {
      if (args === null || args === undefined || typeof args !== "object" || Array.isArray(args)) {
        throw ToolError.validation("Expected args to be an object");
      }

      const required = this.inputSchema["required"];
      if (Array.isArray(required)) {
        const argsRecord = args as Record<string, unknown>;
        for (const field of required) {
          if (typeof field === "string" && !(field in argsRecord)) {
            throw ToolError.missingField(field);
          }
        }
      }
    }
  }
}

// ---------------------------------------------------------------------------
// ToolResult
// ---------------------------------------------------------------------------

export class ToolResult {
  readonly content: ToolContent[];
  readonly isError: boolean;
  readonly citation?: string;
  readonly injectToContext: boolean;
  readonly durationMs?: number;

  constructor(
    content: ToolContent[],
    isError: boolean,
    citation?: string,
    injectToContext: boolean = false,
    durationMs?: number,
  ) {
    this.content = content;
    this.isError = isError;
    this.citation = citation;
    this.injectToContext = injectToContext;
    this.durationMs = durationMs;
  }

  // --- Static factories ---

  static text(text: string): ToolResult {
    return new ToolResult([{ type: "text", data: text }], false);
  }

  static json(data: unknown): ToolResult {
    return new ToolResult([{ type: "json", data }], false);
  }

  static error(message: string): ToolResult {
    return new ToolResult([{ type: "text", data: message }], true);
  }

  // --- Builder methods (immutable) ---

  withCitation(citation: string): ToolResult {
    return new ToolResult(this.content, this.isError, citation, this.injectToContext, this.durationMs);
  }

  withInject(inject: boolean): ToolResult {
    return new ToolResult(this.content, this.isError, this.citation, inject, this.durationMs);
  }

  withDuration(ms: number): ToolResult {
    return new ToolResult(this.content, this.isError, this.citation, this.injectToContext, ms);
  }

  // --- Helpers ---

  /**
   * Returns the first text content value, or undefined if none exists.
   */
  asText(): string | undefined {
    for (const c of this.content) {
      if (c.type === "text") {
        return c.data;
      }
    }
    return undefined;
  }
}

// ---------------------------------------------------------------------------
// ToolContext
// ---------------------------------------------------------------------------

export class ToolContext {
  readonly callerId: string;
  readonly platform: string;
  /** Working directory for this call. File tools resolve relative paths against this. */
  readonly cwd?: string;
  readonly extra: Record<string, unknown>;

  constructor(callerId: string, platform: string, extra: Record<string, unknown> = {}, cwd?: string) {
    this.callerId = callerId;
    this.platform = platform;
    this.cwd = cwd;
    this.extra = extra;
  }

  static create(callerId: string, platform: string): ToolContext {
    return new ToolContext(callerId, platform);
  }

  /** Returns a new ToolContext with cwd set. */
  withCwd(cwd: string): ToolContext {
    return new ToolContext(this.callerId, this.platform, this.extra, cwd);
  }

  /**
   * Returns a new ToolContext with the given key-value pair added to extra.
   */
  with(key: string, value: unknown): ToolContext {
    return new ToolContext(this.callerId, this.platform, { ...this.extra, [key]: value }, this.cwd);
  }

  getStr(key: string): string | undefined {
    const v = this.extra[key];
    return typeof v === "string" ? v : undefined;
  }

  getNum(key: string): number | undefined {
    const v = this.extra[key];
    return typeof v === "number" ? v : undefined;
  }

  getBool(key: string): boolean | undefined {
    const v = this.extra[key];
    return typeof v === "boolean" ? v : undefined;
  }
}

// ---------------------------------------------------------------------------
// Tool interface
// ---------------------------------------------------------------------------

export interface Tool {
  def(): ToolDef;
  call(args: unknown, ctx: ToolContext): Promise<ToolResult>;
}
